### PR TITLE
IVR split viewers: tests

### PR DESCRIPTION
**Pull request by davidpmccormick**
For #12985

## What does this change?

Step 5 of the `MainViewer.tsx` split, following the `VirtualizedImageViewer` /
`PaginatedItemViewer` split in #13304 (since merged to `main`).

Adds unit tests for the components created by the split, plus a matching set for
the legacy implementation they were split out of.

Under `refactored/`:
- **`MainViewer.test.tsx`** (3) — mocks both child viewers as sentinels and asserts pure routing: image-only context renders `VirtualizedImageViewer`, mixed context renders `PaginatedItemViewer`.
- **`VirtualizedImageViewer.test.tsx`** (6) — the fixed-list container and virtualized image items; the initial scroll-to-canvas-on-mount maths, including the landscape-aspect-ratio centring in `scrollViewer`; scroll-driven control hiding/restoring; the `firstItemIsRestricted` margin; and the search-hit highlight overlay.
- **`PaginatedItemViewer.test.tsx`** (7) — current item rendering, the `setShowFullscreenControl(false)` effect, the canvas the query points at, a PDF-from-originals item, a born-digital download link, `externalAccessService` propagation, and the no-resolvable-canvas edge case.

Under `legacy/`:
- **`MainViewer.test.tsx`** (13) — mirrors the two files above test for test. Legacy `MainViewer` is both viewers in one component, so they sit in two describes selected by `hasOnlyRenderableImages`. The shared test bodies are line-for-line identical to their refactored twins, so a diff between the files shows only what the split actually changed.

This addresses the review point about only covering one side: the claim that the
split preserved behaviour is now checked against both implementations rather than
asserted. Two differences remain deliberately — the empty-render test asserts
against `main-viewer` in legacy and the whole container in refactored (legacy
always renders `MainViewerContainer`; the refactored viewer returns `null`), and
`refactored/MainViewer.test.tsx` has no legacy counterpart, since legacy has no
child components to route between.

The refactored tests use the established #13296 pattern: `jest.mock`
`useFeatureFlags` -> `{ itemViewerRefactor: true }` and `renderWithContext(...,
{ useRefactoredContext: true })`. The legacy file does neither, so
`useItemViewerContext` resolves to the legacy context.

### Production code changes

Small, and in service of the tests:
- `refactored/PaginatedItemViewer.tsx` now returns `null` when there's no resolvable canvas, rather than falling out of a `.map()` over an empty array with a redundant `currentCanvas &&` guard inside it. Renders identically.
- Four `data-testid` hooks (`image-item`, `search-term-highlight`) across `legacy/MainViewer.tsx`, `refactored/VirtualizedImageViewer.tsx` and `refactored/VirtualizedImageViewer.SearchHighlights.tsx`, replacing `div > div > div > div` selectors that broke as soon as the DOM shape changed.

`legacy/MainViewer.tsx` is therefore no longer untouched by this branch, as this
description previously claimed — but the two attributes carry no behaviour.

## How to test

- `yarn tsc && yarn lint && yarn test:content` all pass (578 tests, up from 549 on `main`).
- Worth knowing for review: when the legacy tests were first run against `legacy/MainViewer.tsx`, 11 of the 13 passed unmodified. The two that didn't were both DOM-shape assertions rather than behavioural ones, and both were resolved by the `data-testid` hooks above.
- No manual smoke test needed beyond what `playwright/test/view-items-refactored.test.ts` already covers.

## Have we considered potential risks?

Low risk. The only production change with any semantics is the `PaginatedItemViewer`
null return, which renders identically (`[]` and `null` both render nothing) and
is covered by a test on both sides. The `data-testid`s are inert attributes.

Two notes for reviewers:
- Both `refactored/VirtualizedImageViewer.test.tsx` and `legacy/MainViewer.test.tsx` add a local `IntersectionObserver` stub, since jsdom doesn't implement it and these are the first tests to mount a canvas with real `painting` content through the image-scroll-to-url-update path.
- The legacy tests are duplication with a known expiry: they should be deleted in the same commit that deletes `legacy/`. Worth treating that as deliberate and time-boxed rather than something to reconcile later.

### CODE REVIEW

**Review by rcantin-w — COMMENTED** [review]
We missed a big one! Should we be adding these tests to `legacy` as well or are we fully satisfied that what we do won't touch it? These tests were added _after_ our refactor, so do they assume behaviour that's potentially already changed?

File: content/webapp/views/pages/works/work/IIIFViewer/refactored/PaginatedItemViewer.test.tsx [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+  });
+
+  it('hides the fullscreen control, since the paginated viewer never supports it', () => {
+    const setShowFullscreenControl = jest.fn();
+    renderViewer(setShowFullscreenControl);
+
+    expect(setShowFullscreenControl).toHaveBeenCalledWith(false);
+  });
+
+  it('shows the item for the current canvas when paginating to a later canvas', () => {
```
Comment (rcantin-w) [inline content/webapp/views/pages/works/work/IIIFViewer/refactored/PaginatedItemViewer.test.tsx]: This description confounds me? It shows the current canvas when paginating to a different one..?

File: content/webapp/views/pages/works/work/IIIFViewer/refactored/PaginatedItemViewer.test.tsx [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+
+    expect(
+      screen.getByRole('heading', { name: 'Restricted access notice' })
+    ).toBeInTheDocument();
+  });
+
+  it('renders nothing when the current canvas cannot be resolved', () => {
+    // PaginatedItemViewer no longer owns a wrapping container element (that
+    // now lives in the MainViewer router), so an empty render is a bare
+    // fragment - nothing to query for, just an empty container.
```
Comment (rcantin-w) [inline content/webapp/views/pages/works/work/IIIFViewer/refactored/PaginatedItemViewer.test.tsx]: Should this be?

File: content/webapp/views/pages/works/work/IIIFViewer/refactored/VirtualizedImageViewer.test.tsx [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+window.IntersectionObserver =
+  MockIntersectionObserver as unknown as typeof IntersectionObserver;
+
+const renderViewer = (canvases: TransformedCanvas[] = [createMockCanvas()]) =>
+  renderWithContext(<VirtualizedImageViewer />, {
+    contextProps: { transformedManifest: createMockManifest({ canvases }) },
+    useRefactoredContext: true,
+  });
+
+describe('VirtualizedImageViewer', () => {
```
[the remark this answers is not in the document]
Comment (rcantin-w) [reply content/webapp/views/pages/works/work/IIIFViewer/refactored/VirtualizedImageViewer.test.tsx:50]: Sounds good?

File: content/webapp/views/pages/works/work/IIIFViewer/refactored/VirtualizedImageViewer.test.tsx [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+  it('renders the fixed-list scroll container sized to the main area', () => {
+    // The MainViewerContainer/data-testid="main-viewer" wrapper now lives in
+    // the MainViewer router, not here - see MainViewer.test.tsx for that.
```
Comment (rcantin-w) [inline content/webapp/views/pages/works/work/IIIFViewer/refactored/VirtualizedImageViewer.test.tsx]: When we review which tests we keep, we should review which comments we keep too (this one makes sense now but won't after the work is done)

File: content/webapp/views/pages/works/work/IIIFViewer/refactored/VirtualizedImageViewer.test.tsx [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+            },
+          ],
+          hits: [],
+        },
+      },
+      useRefactoredContext: true,
+    });
+
+    // The highlight overlay renders as an extra sibling alongside the item
+    // wrapper, once the image position effect has run. RTL's own mounting
```
Comment (rcantin-w) [inline content/webapp/views/pages/works/work/IIIFViewer/refactored/VirtualizedImageViewer.test.tsx]:
```suggestion
    // wrapper, once the image position effect has run. `react-testing-library` own mounting
```

(My brain went "right to left")

File: content/webapp/views/pages/works/work/IIIFViewer/refactored/VirtualizedImageViewer.test.tsx [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+    // div counts as one ancestor level: container(RTL) > scroll > sizer > row.
+    const row = container.querySelector('div > div > div > div');
```
Comment (rcantin-w) [inline content/webapp/views/pages/works/work/IIIFViewer/refactored/VirtualizedImageViewer.test.tsx]: Wonder if we could have a better selector here (we use it twice), maybe even just adding a `data-testid` (usually a last resource but hey ho https://playwright.dev/docs/locators#locating-elements)

I could do a pass that ensures it's fully aligned to our usual testing approach?

File: content/webapp/views/pages/works/work/IIIFViewer/refactored/VirtualizedImageViewer.test.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (davidpmccormick) [reply content/webapp/views/pages/works/work/IIIFViewer/refactored/VirtualizedImageViewer.test.tsx:50]: https://github.com/wellcomecollection/wellcomecollection.org/pull/13305/changes/034d2b9872f2ab0d5b59bdbfc92be8a56b12b4b9

File: content/webapp/views/pages/works/work/IIIFViewer/refactored/PaginatedItemViewer.test.tsx (hunk 1 above)
Comment (davidpmccormick) [reply content/webapp/views/pages/works/work/IIIFViewer/refactored/PaginatedItemViewer.test.tsx]: yeah, it's trying to say it keeps the current canvas number up to date, or similar. Needs a rewrite for sure!

Comment (davidpmccormick) [reply content/webapp/views/pages/works/work/IIIFViewer/refactored/PaginatedItemViewer.test.tsx]: https://github.com/wellcomecollection/wellcomecollection.org/pull/13305/changes/4e380632428dc87820b18e4f19230c2a3008fb77

File: content/webapp/views/pages/works/work/IIIFViewer/refactored/PaginatedItemViewer.test.tsx (hunk 2 above)
Comment (rcantin-w) [reply content/webapp/views/pages/works/work/IIIFViewer/refactored/PaginatedItemViewer.test.tsx]: https://github.com/wellcomecollection/wellcomecollection.org/pull/13305/changes/9337b3eb2c94152fa5c458da90ecc4ccc19c4d8d

File: content/webapp/views/pages/works/work/IIIFViewer/refactored/VirtualizedImageViewer.test.tsx (hunk 6 above)
Comment (davidpmccormick) [reply content/webapp/views/pages/works/work/IIIFViewer/refactored/VirtualizedImageViewer.test.tsx]: Fixed with data-testid

**Comment by davidpmccormick** [comment]
> We missed a big one! Should we be adding these tests to legacy as well or are we fully satisfied that what we do won't touch it? These tests were added after our refactor, so do they assume behaviour that's potentially already changed?

We spoke, and agreed we should have equivalent tests for both.

https://github.com/wellcomecollection/wellcomecollection.org/pull/13305/changes/b77196bbd143114869b1409a3157c4ebc2c96fa9

File: content/webapp/views/pages/works/work/IIIFViewer/refactored/VirtualizedImageViewer.test.tsx (hunk 4 above)
Comment (davidpmccormick) [reply content/webapp/views/pages/works/work/IIIFViewer/refactored/VirtualizedImageViewer.test.tsx]: Added to #13273 description

File: content/webapp/test/fixtures/iiif/transformed-manifest.ts
Comment (davidpmccormick) [inline content/webapp/test/fixtures/iiif/transformed-manifest.ts:1]: These factories remove some duplication in the test files

File: content/webapp/views/pages/works/work/IIIFViewer/legacy/MainViewer.test.tsx [hunk 7]
```diff
@@ -0,0 +1,409 @@
...
+      // quietly passing.
+      //
+      // mainAreaWidth/itemSize default to 1000, mainAreaHeight to 500.
+      // renderedHeight = 1000 * (1000/2000) * 0.8 = 400
+      // heightOfPreviousItems = 1 * 1000 = 1000
+      // distanceToScroll = 1000 + (1000 - 400) / 2 = 1300
+      expect(getScrollContainer(container).scrollTop).toBe(1300);
+    });
+
+    it('scrolls to the top of the current canvas on mount when it is portrait', () => {
```
Comment (davidpmccormick) [inline content/webapp/views/pages/works/work/IIIFViewer/legacy/MainViewer.test.tsx:135]: Claude caught that we weren't testing this portrait path even though it is the more common

File: content/webapp/views/pages/works/work/IIIFViewer/legacy/MainViewer.test.tsx [hunk 8]
```diff
@@ -0,0 +1,409 @@
...
+        query: createMockQuery({ canvas: 2 }),
+      });
+
+      // Portrait canvases skip the centring maths above entirely: scrollViewer
+      // calls scrollToItem(1, 'start'), putting the top of the second item at the
+      // top of the viewport - 1 * itemSize = 1000.
+      expect(getScrollContainer(container).scrollTop).toBe(1000);
+    });
+
+    it('skips the canvas-change scroll when shouldScrollToCanvas is false', () => {
```
Comment (davidpmccormick) [inline content/webapp/views/pages/works/work/IIIFViewer/legacy/MainViewer.test.tsx:160]: The `shouldScrollToCanvas` behaviour was previously untested, so this is new. There's potentially a bug here it seems:

> **What I found.** Two separate bits of code can move the scroll position:
> 
> 1. The effect that watches the page number. This one checks the flag.
> 2. A handler that runs once on first render, to jump you to whichever page the URL asked for when you opened the viewer. This one does not check the flag — it just scrolls.
> 
> So with the flag set to false, the viewer sits still at first, and then half a second later jumps anyway, because the second bit of code fires on a 500ms delay and never consulted the flag.

Since these are supposed to check that the behaviour stays the same between `/legacy` and `/refactored`, I don't think this is the right time to fix it, but I'll add it to our follow up ticket

**Review by rcantin-w — APPROVED** [review]
